### PR TITLE
coll: refactor tsp based infrastructure

### PR DESCRIPTION
**Pull request by hzhou**
Note: there are much comment in the individual commit messages.

### CODE REVIEW

**Comment by hzhou** [comment]
@wesbland @raffenet I'll fix the formatting (and compilation) issues next week. But first let's focus on the infrastructures. Please send feedbacks as I continue working on this.

File: src/mpi/coll/iallgather/iallgather_tsp_brucks_algos.h [hunk 1]
```diff
@@ -67,12 +67,14 @@ MPIR_TSP_Iallgather_sched_intra_brucks(const void *sendbuf, int sendcount,
     /* get datatype info of sendtype and recvtype */
     MPIR_Datatype_get_size_macro(sendtype, sendtype_size);
     MPIR_Datatype_get_extent_macro(sendtype, sendtype_extent);
-    MPIR_Type_get_true_extent_impl(sendtype, &sendtype_lb, &sendtype_true_extent);
+    MPIR_Type_get_true_extent_impl(sendtype, (MPI_Aint *) & sendtype_lb,
```
Comment (raffenet) [inline src/mpi/coll/iallgather/iallgather_tsp_brucks_algos.h]: These additional casts (which is presume are squashing warnings?) don't belong with this argument re-ordering, IMO.

**Comment by raffenet** [comment]
cc: @pmodels/mpich-intel see my review comment above.

**Comment by wesbland** [comment]
I think the better reviewer here is @zhenggb72 since he's running the collectives efforts. I'm going to remove myself and add him.

**Comment by hzhou** [comment]
This PR is not really as big as it appears. Most of the changes are template stuff gets repeated. For reference, here is the template/mydef source that I use to generate them:
    https://github.com/hzhou/mpich_study/blob/master/coll/mpir_coll_h.def
    https://github.com/hzhou/mpich_study/blob/master/coll/tsp_algos.def
and 
    https://github.com/hzhou/mpich_study/blob/master/coll/coll_common.def

**Comment by hzhou** [comment]
I just realized I made a typo in my private branch name. It was supposed to be `1901_coll_refactor`. Probably nobody cares, but just in case if some one wonders :)

**Comment by rkalidas** [comment]
We have lot of algorithm PRs internally and if this PR gets merged we will have a tough time during our weekly rebase. I suggest we hold off on this until we create public PRs for all algorithms.

**Comment by zhenggb72** [comment]
This PR looks ok to me, and the documentation looks good except that I don't know if it is a good idea to explain all coll algorithms in a single document (we can have comments in the algorithm code itself.).

Also, it seems that this PR will create many conflicts (due to removing some files) with our PRs in our integration branch. We have at least 10 PRs need to be submitted/reviewed, and that is our current main focus to clear the backlog of our local PRs. I would prefer we merge this PR after our major PRs are reviewed and merged.

**Comment by hzhou** [comment]
I agree that other colll PRs can be reviewed and committed first. However, I would like this PR to be reviewed first. It is important to reach an agreement and understanding. As long as we are in agreement, I could simply ignore the infrastructure part of the other PRs and only focus on the `..._algos.h` and tests.

**Comment by hzhou** [comment]
@zhenggb72 The point of notes is a place for developers writing down understandings (that took them a while to work out and would like to save others time) and communicate to each other. For it to work, I intend them to be very relaxed -- any one can write down or change or delete whatever notes they deem appropriate. Since they are in the repository, they gets reviewed -- which is the main point.

If we start to debate on what is appropriate content and place, due to the subjective nature, they are difficult to resolve. As a consequence, these high-level overview notes don't get to written down anywhere and every developer has to "reverse-engineer" it every time.

**Comment by wesbland** [comment]
@hzhou - The list of PRs is here: https://github.com/pmodels/mpich/pull/3490/files#diff-9f7195eef0002e1672710a84c85286ee in the "Collective Algorithms" column.

@zhenggb72 - Do we have any algorithms internally that still need to have external PRs made? I think the answer is yes.

**Comment by hzhou** [comment]
> @hzhou - The list of PRs is here: https://github.com/pmodels/mpich/pull/3490/files#diff-9f7195eef0002e1672710a84c85286ee in the "Collective Algorithms" column.

Wrong link?

**Comment by hzhou** [comment]
https://github.com/pmodels/mpich/projects/1  -- "Collective" column. 

We'll merge this PR once we clear that column.

**Comment by wesbland** [comment]
> > @hzhou - The list of PRs is here: https://github.com/pmodels/mpich/pull/3490/files#diff-9f7195eef0002e1672710a84c85286ee in the "Collective Algorithms" column.
> 
> Wrong link?

Oops. 😄 https://github.com/pmodels/mpich/projects/1

Don't you want to review shared memory collectives too?

**Comment by hzhou** [comment]
> Don't you want to review shared memory collectives too?

Maybe, depend on how related they are. One block at a time :)

**Comment by jain-surabhi-23** [comment]
> > Don't you want to review shared memory collectives too?
> 
> Maybe, depend on how related they are. One block at a time :)

Shared Memory Collectives are not related to all this. It has lot of documentation as well, including a paper :)

**Comment by wesbland** [comment]
@jain-surabhi-23 - That was a joke because I accidentally posted a link to that PR. Not that I want to tie these together.

**Comment by jain-surabhi-23** [comment]
> @jain-surabhi-23 - That was a joke because I accidentally posted a link to that PR. Not that I want to tie these together.

Yes. That I got. Just wanted to answer @hzhou's question.

**Comment by pavanbalaji** [comment]
Refactoring patch.  Should be delayed till the next major release.

**Pull request closed without merging**